src/platform/forms-system/src/js/state/helpers.js

```diff
@@ -358,28 +358,84 @@ export function updateSchemaFromUiSchema(
   return currentSchema;
 }
 
+// These are managed by other flows, or already handle dynamic behavior,
+// so we don’t want to allow these to be updated via updateUiSchema
+const DISALLOWED_UPDATE_UI_SCHEMA_PROPS = {
+  'ui:field': true,
+  'ui:required': true,
+  'ui:reviewField': true,
+  'ui:validations': true,
+  hideIf: true,
+  updateSchema: true,
+  updateUiSchema: true,
+  replaceSchema: true,
+  viewComponent: true,
+  viewField: true,
+};
+
+/**
+ * Merges a partial uiSchema into an existing uiSchema, but only if
+ * the new uiSchema is different from the existing one.
+ * Does not support arrays.
+ * @param {UISchemaOptions} uiSchema - The existing uiSchema
+ * @param {UISchemaOptions} newUiSchema - A partial uiSchema to merge into the existing one
+ * @returns {UISchemaOptions} Merged or original uiSchema
+ */
+function mergeUiSchemasIfDifferent(uiSchema, newUiSchema) {
+  if (!newUiSchema) {
+    return uiSchema;
+  }
+
+  let isDifferent = false;
+  const updatedUiSchema = { ...uiSchema };
+
+  Object.entries(newUiSchema).forEach(([prop, value]) => {
+    let newValue = value;
+    if (DISALLOWED_UPDATE_UI_SCHEMA_PROPS[prop]) {
+      throw new Error(
+        `Cannot update uiSchema property '${prop}' using updateUiSchema.`,
+      );
+    }
+    if (value && typeof value === 'object' && !Array.isArray(value)) {
+      newValue = mergeUiSchemasIfDifferent(uiSchema[prop], newValue);
+    }
+
+    if (updatedUiSchema[prop] !== newValue) {
+      updatedUiSchema[prop] = newValue;
+      isDifferent = true;
+    }
+  });
+
+  return isDifferent ? updatedUiSchema : uiSchema;
+}
+
 /**
  * A helper that returns a new uiSchema based on the input uiSchema and
  * formData. Only updates and returns a uiSchema object for the field that has
  * an `updateUiSchema` callback defined in its `ui:options`.
  *
- * Note that this helper is _not_ called as part of the normal data-update flow
- * in a standard form config-powered benefit application. Perhaps it could be
- * called as part of updateSchemaAndData() (also in this file) but we'd have to
- * investigate any possible side effects of updating that function to add a
- * uiSchema prop in the object it returns.
- *
- * @param {Object} uiSchema - The uiSchema to update
+ * @param {SchemaOptions} schema - The schema
+ * @param {UISchemaOptions} uiSchema - The uiSchema to update
  * @param {Object} formData - The form data to based uiSchema updates on
- * @returns {Object} The new uiSchema object
+ * @returns {UISchemaOptions} The new uiSchema object
  */
-export function updateUiSchema(uiSchema, formData) {
+export function updateUiSchema(schema, uiSchema, formData) {
```

Comment:
note: this existing function was not part of the normal formData update flow, and was only used as a one off in two other places in the code. I updated it so that now it IS part of the normal formData update flow, and now also allows for deeply nested options ('ui:options')